### PR TITLE
impl From<Indeterminate<Error>> for std::io::Error

### PR DESCRIPTION
**Pull request by tylerfanelli**
Implementing #31 allowing more convenience for callers. If Indeterminate<Error> is a known IoError, the io::Error can simply be returned for context. If not, io::Error::Other is returned.

### CODE REVIEW

**Comment by connorkuehl** [comment]
@tfanelli-rh Please fix the `cargo clippy` errors. Don't sweat the failing tests, that's an infrastructure thing, not a your-PR-thing.

File: src/util/cached_chain.rs [hunk 1]
```diff
@@ -55,8 +55,7 @@ pub fn sys() -> Option<PathBuf> {
 pub fn path() -> Vec<PathBuf> {
     vec![env_var(), home(), sys()]
         .into_iter()
-        .filter(|p| p.is_some())
-        .map(|p| p.unwrap())
+        .flatten()
```
Comment (connorkuehl) [inline src/util/cached_chain.rs:58]: This should be in a separate commit because it is a logical change of its own and not related to the other change in this commit.

File: src/firmware/mod.rs [hunk 2]
```diff
@@ -128,6 +128,17 @@ impl From<std::io::Error> for Indeterminate<Error> {
...
 }
 
+impl From<Indeterminate<Error>> for std::io::Error {
+    #[inline]
+    fn from(error: Indeterminate<Error>) -> std::io::Error {
+        if let Indeterminate::Known(Error::IoError(io_err)) = error {
+            return io_err;
+        }
+
+        std::io::ErrorKind::Other.into()
```
Comment (connorkuehl) [inline src/firmware/mod.rs]: Would you consider `impl std::error::Error for Error`? I think in that case you could take advantage of the `std::io::Error::new` constructor function as it takes a std::io::ErrorKind and any type that implements std::error::Error as a payload. Then you could just unpack the `Indeterminate::Known(blah)` and return `std::io::Error::new(ErrorKind::Other, blah)` Otherwise if it's Unknown you can use some other `std::io::ErrorKind`.

**Review by connorkuehl — CHANGES_REQUESTED** [review]
Please improve the commit message.

You can't `impl Indeterminate for std::io::Error` because `Indeterminate` is not a trait.

What you're doing here is `impl From<Indeterminate<Error>> for std::io::Error`.

I'd also recommend putting part of your justification in your PR comment in the commit as well. Something about how it is convenient for callers would be good justification, in fact, we'd like to consume this downstream in `sevctl` to avoid having to map these errors into io::Errors.

And don't forget to sign your commits. I sign my commits by putting this at the end of my commit message:

```
Signed-off-by: Connor Kuehl <ckuehl@redhat.com>
```

You can tell `git` to automatically sign your commits with `git commit -s`.

File: src/firmware/mod.rs [hunk 3]
```diff
@@ -114,6 +114,56 @@ pub enum Error {
...
 }
 
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let err_description = match self {
+            Error::IoError(_) => "I/O Error",
+            Error::InvalidPlatformState => "Invalid platform state",
+            Error::InvalidGuestState => "Invalid guest state",
+            Error::InvalidConfig => "Platform configuration invalid",
+            Error::InvalidLen => "Memory buffer to small",
```
Comment (connorkuehl) [inline src/firmware/mod.rs]: typo: "to" -> "too"

File: src/firmware/mod.rs [hunk 4]
```diff
@@ -114,6 +114,56 @@ pub enum Error {
...
+            Error::InvalidCertificate => "Invalid certificate",
+            Error::PolicyFailure => "Policy failure",
+            Error::Inactive => "Guest is inactive",
+            Error::InvalidAddress => "Provided address is invalid",
+            Error::BadSignature => "Provided signature is invalid",
+            Error::BadMeasurement => "Provided measurement is invalid",
+            Error::AsidOwned => "ASID is already owned",
+            Error::InvalidAsid => "ASID is invalid",
+            Error::WbinvdRequired => "WBINVD instruction required",
+            Error::DfFlushRequired => "`DF_FLUSH` invocation required",
```
Comment (connorkuehl) [inline src/firmware/mod.rs]: Why the backticks?

File: src/firmware/mod.rs [hunk 5]
```diff
@@ -114,6 +114,56 @@ pub enum Error {
...
+            Error::InvalidGuest => "Guest handle is invalid",
+            Error::InvalidCommand => "Issued command is invalid",
+            Error::Active => "Guest is active",
+            Error::HardwarePlatform => {
+                "Hardware condition occured, safe to re-allocate parameter buffers"
+            }
+            Error::HardwareUnsafe => {
+                "Hardware condition occured, unsafe to re-allocate parameter buffers"
+            }
```
Comment (connorkuehl) [inline src/firmware/mod.rs:145]: Just curious why the curly braces are used here but not for the others.

File: src/firmware/mod.rs [hunk 6]
```diff
@@ -114,6 +114,56 @@ pub enum Error {
...
+}
+
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        if let Error::IoError(io_err) = self {
+            return Some(io_err);
+        }
+
+        None
+    }
```
Comment (connorkuehl) [inline src/firmware/mod.rs]:
```suggestion
        match self {
            Error::IoError(e) => Some(e),
            _ => None,
        }
```

File: src/firmware/mod.rs [hunk 7]
```diff
@@ -128,6 +178,17 @@ impl From<std::io::Error> for Indeterminate<Error> {
...
 }
 
+impl From<Indeterminate<Error>> for std::io::Error {
+    #[inline]
+    fn from(indeterminate: Indeterminate<Error>) -> std::io::Error {
+        if let Indeterminate::Known(known_err) = indeterminate {
+            return std::io::Error::new(std::io::ErrorKind::Other, known_err);
+        }
+
+        std::io::ErrorKind::Other.into()
```
Comment (connorkuehl) [inline src/firmware/mod.rs]:
```suggestion
        match indeterminate {
            Indeterminate::Known(e) => std::io::Error::new(std::io::ErrorKind::Other, e),
            Indeterminate::Unknown => std::io::Error::new(std::io::ErrorKind::Other, "unknown SEV error"),
        }
\```

**Review by connorkuehl — CHANGES_REQUESTED** [review]
Looks good overall! I added a couple of alternatives using `match` statements, which I find appealing because they seem more expressive and less imperative looking.

Lastly, you might want to consider putting some `use` statements to work so there's less typing involved. Such as `use std::{error, io};` or something like that. Then you can refer to `io::Error`, `io::ErrorKind`, etc, instead of their absolute paths.

File: src/firmware/mod.rs (hunk 5 above)
Comment (tylerfanelli) [reply src/firmware/mod.rs:145]: `cargo fmt` placed those there, I assume because the original draft exceeded the 80 character line limit the Rust style suggests. If you think they're better off removed I can position everything on one line.

File: src/firmware/mod.rs (hunk 4 above)
Comment (tylerfanelli) [reply src/firmware/mod.rs]: There's a description for `DfFlushRequired` in the `enum Error` that uses the backticks. I'll remove them.

File: src/firmware/mod.rs (hunk 5 above)
Comment (connorkuehl) [reply src/firmware/mod.rs:145]: No, I prefer whatever cargo fmt wants.

**Review by connorkuehl — APPROVED** [review]
Thanks! :slightly_smiling_face:

**Review by haraldh — APPROVED** [review]
